File: backend-python/app/modules/system/controller/menu_controller.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession
from app.common.api_response import ApiResponse, success
from app.modules.system.dto.menu_request import MenuSaveRequest
from app.modules.system.dto.user_request import StatusUpdateRequest
from app.modules.system.router import get_db, require_permission
from app.modules.system.service.menu_service import MenuService
router=APIRouter(tags=["menu"])
def service(db:AsyncSession=Depends(get_db))->MenuService:return MenuService(db)
@router.get("/menu",response_model=ApiResponse)
async def tree(_:dict=Depends(require_permission("system:menu:list")),svc:MenuService=Depends(service))->ApiResponse:return success(await svc.all_tree())
@router.post("/menu",response_model=ApiResponse)
async def create(body:MenuSaveRequest,_:dict=Depends(require_permission("system:menu:create")),svc:MenuService=Depends(service))->ApiResponse:return success(await svc.save(None,body))
@router.put("/menu/{menu_id}",response_model=ApiResponse)
async def update(menu_id:int,body:MenuSaveRequest,_:dict=Depends(require_permission("system:menu:update")),svc:MenuService=Depends(service))->ApiResponse:await svc.save(menu_id,body);return success()
@router.delete("/menu/{menu_id}",response_model=ApiResponse)
async def delete(menu_id:int,_:dict=Depends(require_permission("system:menu:delete")),svc:MenuService=Depends(service))->ApiResponse:await svc.delete(menu_id);return success()
@router.put("/menu/{menu_id}/status",response_model=ApiResponse)
async def status(menu_id:int,body:StatusUpdateRequest,_:dict=Depends(require_permission("system:menu:status")),svc:MenuService=Depends(service))->ApiResponse:await svc.update_status(menu_id,body.status);return success()

File: backend-rust/src/modules/system/controller/menu_controller.rs
use super::{ok, ok_null};
use crate::common::{ApiResponse, AppError, I64String};
use crate::modules::system::AppState;
use crate::modules::system::dto::menu_request::MenuSaveRequest;
use crate::modules::system::dto::status_update_request::StatusUpdateRequest;
use crate::modules::system::service::menu_service;
use crate::modules::system::vo::menu_tree_vo::MenuTreeVo;
use axum::Json;
use axum::extract::{Path, State};

pub async fn list(
    State(state): State<AppState>,
) -> Result<Json<ApiResponse<Vec<MenuTreeVo>>>, AppError> {
    Ok(ok(menu_service::list(state.database.as_ref()).await?))
}

pub async fn create(
    State(state): State<AppState>,
    Json(request): Json<MenuSaveRequest>,
) -> Result<Json<ApiResponse<I64String>>, AppError> {
    Ok(ok(
        menu_service::create(state.database.as_ref(), request).await?
    ))
}

pub async fn update(
    State(state): State<AppState>,
    Path(id): Path<i64>,
    Json(request): Json<MenuSaveRequest>,
) -> Result<Json<ApiResponse<()>>, AppError> {
    menu_service::update(state.database.as_ref(), id, request).await?;
    Ok(ok_null())
}

pub async fn delete(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> Result<Json<ApiResponse<()>>, AppError> {
    menu_service::delete(state.database.as_ref(), id).await?;
    Ok(ok_null())
}

pub async fn status(
    State(state): State<AppState>,
    Path(id): Path<i64>,
    Json(request): Json<StatusUpdateRequest>,
) -> Result<Json<ApiResponse<()>>, AppError> {
    menu_service::update_status(state.database.as_ref(), id, request.status).await?;
    Ok(ok_null())
}
